eoe/man/man1/attr.1
.nr X
.if \nX=0 .ds x} attr 1 "Essential Utilities" "\&"
.TH \*(x}
.SH NAME
attr \- manipulate Extended Attributes on filesystem objects
.SH SYNOPSIS
.nf
\f3attr\f1 [ \f3\-LRq\f1 ] \f3\-s attrname\f1 [ \f3\-V attrvalue\f1 ] \c
\f3pathname\f1
.sp .8v
\f3attr\f1 [ \f3\-LRq\f1 ] \f3\-g attrname pathname\f1
.sp .8v
\f3attr\f1 [ \f3\-LRq\f1 ] \f3\-r attrname pathname\f1
.sp .8v
\f3attr\f1 [ \f3\-LRq\f1 ] \f3\-l pathname\f1
.sp .8v
.fi
.SH OVERVIEW
Extended Attributes implement the ability for a user to attach
name/value pairs to objects within the filesystem.
They are currently only supported in XFS filesystems.
.P
They could be used to store meta-information about the file.
For example "character-set=kanji" could tell a document browser to
use the Kanji character set when displaying that document
and "thumbnail=..." could provide a reduced resolution overview of a
high resolution graphic image.
.P
The
.I names
can be up to 256 bytes in length, terminated by the first 0 byte.
The intent is that they be printable ASCII (or other character set)
names for the attribute.
.P
The
.I values
can be up to 256KB of arbitrary binary data.
.P
Attributes can be attached to all types of inodes:
regular files, directories, symbolic links, device nodes, etc.
.P
There are 2 disjoint attribute name spaces associated with every
filesystem object.
They are the
.B root
and
.B user
address spaces.
The
.B root
address space is accessable only to the superuser,
and then only by specifying a flag argument to the function call.
Other users will not see or be able to modify attributes in the
.B root
address space.
The
.B user
address space is protected by the normal file permissions mechanism,
so the owner of the file can decide who is able to see and/or modify
the value of attributes on any particular file.
.P
Attributes are currently supported only in the XFS filesystem type.
.SH DESCRIPTION
The
.I attr
utility allows the manipulation of Extended Attributes associated with
filesystem objects from within shell scripts.
.PP
There are four main operations that
.I attr
can perform:
.TP
.B GET
The
.B \-g attrname
option tells
.I attr
to search the named object and print (to \f4stdout\fP) the value
associated with that attribute name.
With the
.B \-q
flag, \f4stdout\fP will be exactly and only the value of the attribute,
suitable for storage directly into a file or processing via a piped command.
.TP
.B LIST
The
.B \-l
option tells
.I attr
to list the names of all the attributes that are associated with the object,
and the number of bytes in the value of each of those attributes.
With the
.B \-q
flag, \f4stdout\fP will be a simple list of only the attribute names,
one per line, suitable for input into a script.
.TP
.B REMOVE
The
.B \-r attrname
option tells
.I attr
to remove an attribute with the given name from the object if the
attribute exists.
There is no output on sucessful completion.
.TP
.B SET/CREATE
The
.B \-s attrname
option tells
.I attr
to set the named attribute of the object to the value read from \f4stdin\fP.
If an attribute with that name already exists,
its value will be replaced with this one.
If an attribute with that name does not already exist,
one will be created with this value.
With the
.B \-V attrvalue
flag, the attribute will be set to have a value of
.B attrvalue
and \f4stdin\fP will not be read.
With the
.B \-q
flag, \f4stdout\fP will not be used.
Without the
.B \-q
flag, a message showing the attribute name and the entire value
will be printed.
.PP
When the
.B \-L
option is given and the named object is a symbolic link,
operate on the attributes of the object referenced by the symbolic link.
Without this option, operate on the attributes of the symbolic link itself.
.PP
When the
.B \-R
option is given and the process has appropriate privileges,
operate in the
.I root
attribute namespace rather that the
.I USER
attribute namespace.
.PP
When the
.B \-q
option is given
.I attr
will try to keep quiet.
It will output error messages (to \f4stderr\fP)
but will not print status messages (to \f4stdout\fP).
.SH "NOTES"
The standard file interchange/archive programs
.IR tar (1),
.IR cpio (1),
and
.IR bru (1)
will not archive or restore Extended Attributes,
while the
.IR xfsdump (1m)
program will.
.SH "SEE ALSO"
.\"	@(#)attr.1	1.0 of 6/12/95
attr_get(2), attr_getf(2),
attr_list(2), attr_listf(2),
attr_multi(2), attr_multif(2),
attr_remove(2), attr_removef(2),
attr_set(2), attr_setf(2),
xfsdump(1m).


irix/cmd/xfs/db/attr.c
#ident "$Revision: 1.15 $"

#include "versions.h"
#include <sys/param.h>
#include <sys/uuid.h>
#include <assert.h>
#include <stddef.h>
#include <stdio.h>
#include <sys/fs/xfs_types.h>
#include <sys/fs/xfs_inum.h>
#include <sys/fs/xfs_sb.h>
#include <sys/fs/xfs_alloc_btree.h>
#include <sys/fs/xfs_bmap_btree.h>
#include <sys/fs/xfs_dir_sf.h>
#if VERS >= V_654
#include <sys/fs/xfs_dir2_sf.h>
#endif
#include <sys/fs/xfs_attr_sf.h>
#include <sys/fs/xfs_dinode.h>
#include <sys/fs/xfs_da_btree.h>
#include <sys/fs/xfs_attr_leaf.h>
#include "sim.h"
#include "bit.h"
#include "type.h"
#include "faddr.h"
#include "fprint.h"
#include "field.h"
#include "attr.h"
#include "io.h"
#include "data.h"
#include "mount.h"

static int	attr_leaf_entries_count(void *obj, int startoff);
static int	attr_leaf_hdr_count(void *obj, int startoff);
static int	attr_leaf_name_local_count(void *obj, int startoff);
static int	attr_leaf_name_local_name_count(void *obj, int startoff);
static int	attr_leaf_name_local_value_count(void *obj, int startoff);
static int	attr_leaf_name_local_value_offset(void *obj, int startoff,
						  int idx);
static int	attr_leaf_name_remote_count(void *obj, int startoff);
static int	attr_leaf_name_remote_name_count(void *obj, int startoff);
static int	attr_leaf_nvlist_count(void *obj, int startoff);
static int	attr_leaf_nvlist_offset(void *obj, int startoff, int idx);
static int	attr_node_btree_count(void *obj, int startoff);
static int	attr_node_hdr_count(void *obj, int startoff);

const field_t	attr_hfld[] = {
	{ "", FLDT_ATTR, OI(0), C1, 0, TYP_NONE },
	{ NULL }
};

#define	LOFF(f)	bitize(offsetof(xfs_attr_leafblock_t, f))
#define	NOFF(f)	bitize(offsetof(xfs_da_intnode_t, f))
const field_t	attr_flds[] = {
	{ "hdr", FLDT_ATTR_LEAF_HDR, OI(LOFF(hdr)), attr_leaf_hdr_count,
	  FLD_COUNT, TYP_NONE },
	{ "hdr", FLDT_ATTR_NODE_HDR, OI(NOFF(hdr)), attr_node_hdr_count,
	  FLD_COUNT, TYP_NONE },
	{ "entries", FLDT_ATTR_LEAF_ENTRY, OI(LOFF(entries)),
	  attr_leaf_entries_count, FLD_ARRAY|FLD_COUNT, TYP_NONE },
	{ "btree", FLDT_ATTR_NODE_ENTRY, OI(NOFF(btree)), attr_node_btree_count,
	  FLD_ARRAY|FLD_COUNT, TYP_NONE },
	{ "nvlist", FLDT_ATTR_LEAF_NAME, attr_leaf_nvlist_offset,
	  attr_leaf_nvlist_count, FLD_ARRAY|FLD_OFFSET|FLD_COUNT, TYP_NONE },
	{ NULL }
};

#define	BOFF(f)	bitize(offsetof(xfs_da_blkinfo_t, f))
const field_t	attr_blkinfo_flds[] = {
	{ "forw", FLDT_ATTRBLOCK, OI(BOFF(forw)), C1, 0, TYP_ATTR },
	{ "back", FLDT_ATTRBLOCK, OI(BOFF(back)), C1, 0, TYP_ATTR },
	{ "magic", FLDT_UINT16X, OI(BOFF(magic)), C1, 0, TYP_NONE },
#if VERS >= V_65
	{ "pad", FLDT_UINT16X, OI(BOFF(pad)), C1, FLD_SKIPALL, TYP_NONE },
#endif
	{ NULL }
};

#define	LEOFF(f)	bitize(offsetof(xfs_attr_leaf_entry_t, f))
const field_t	attr_leaf_entry_flds[] = {
	{ "hashval", FLDT_UINT32X, OI(LEOFF(hashval)), C1, 0, TYP_NONE },
	{ "nameidx", FLDT_UINT16D, OI(LEOFF(nameidx)), C1, 0, TYP_NONE },
	{ "flags", FLDT_UINT8X, OI(LEOFF(flags)), C1, FLD_SKIPALL, TYP_NONE },
	{ "incomplete", FLDT_UINT1,
	  OI(LEOFF(flags) + bitsz(__uint8_t) - XFS_ATTR_INCOMPLETE_BIT - 1), C1,
	  0, TYP_NONE },
	{ "root", FLDT_UINT1,
	  OI(LEOFF(flags) + bitsz(__uint8_t) - XFS_ATTR_ROOT_BIT - 1), C1, 0,
	  TYP_NONE },
	{ "local", FLDT_UINT1,
	  OI(LEOFF(flags) + bitsz(__uint8_t) - XFS_ATTR_LOCAL_BIT - 1), C1, 0,
	  TYP_NONE },
	{ "pad2", FLDT_UINT8X, OI(LEOFF(pad2)), C1, FLD_SKIPALL, TYP_NONE },
	{ NULL }
};

#define	LHOFF(f)	bitize(offsetof(xfs_attr_leaf_hdr_t, f))
const field_t	attr_leaf_hdr_flds[] = {
	{ "info", FLDT_ATTR_BLKINFO, OI(LHOFF(info)), C1, 0, TYP_NONE },
	{ "count", FLDT_UINT16D, OI(LHOFF(count)), C1, 0, TYP_NONE },
	{ "usedbytes", FLDT_UINT16D, OI(LHOFF(usedbytes)), C1, 0, TYP_NONE },
	{ "firstused", FLDT_UINT16D, OI(LHOFF(firstused)), C1, 0, TYP_NONE },
	{ "holes", FLDT_UINT8D, OI(LHOFF(holes)), C1, 0, TYP_NONE },
	{ "pad1", FLDT_UINT8X, OI(LHOFF(pad1)), C1, FLD_SKIPALL, TYP_NONE },
	{ "freemap", FLDT_ATTR_LEAF_MAP, OI(LHOFF(freemap)),
	  CI(XFS_ATTR_LEAF_MAPSIZE), FLD_ARRAY, TYP_NONE },
	{ NULL }
};

#define	LMOFF(f)	bitize(offsetof(xfs_attr_leaf_map_t, f))
const field_t	attr_leaf_map_flds[] = {
	{ "base", FLDT_UINT16D, OI(LMOFF(base)), C1, 0, TYP_NONE },
	{ "size", FLDT_UINT16D, OI(LMOFF(size)), C1, 0, TYP_NONE },
	{ NULL }
};

#define	LNOFF(f)	bitize(offsetof(xfs_attr_leaf_name_local_t, f))
#define	LVOFF(f)	bitize(offsetof(xfs_attr_leaf_name_remote_t, f))
const field_t	attr_leaf_name_flds[] = {
	{ "valuelen", FLDT_UINT16D, OI(LNOFF(valuelen)),
	  attr_leaf_name_local_count, FLD_COUNT, TYP_NONE },
	{ "namelen", FLDT_UINT8D, OI(LNOFF(namelen)),
	  attr_leaf_name_local_count, FLD_COUNT, TYP_NONE },
	{ "name", FLDT_CHARNS, OI(LNOFF(nameval)),
	  attr_leaf_name_local_name_count, FLD_COUNT, TYP_NONE },
	{ "value", FLDT_CHARNS, attr_leaf_name_local_value_offset,
	  attr_leaf_name_local_value_count, FLD_COUNT|FLD_OFFSET, TYP_NONE },
	{ "valueblk", FLDT_UINT32X, OI(LVOFF(valueblk)),
	  attr_leaf_name_remote_count, FLD_COUNT, TYP_NONE },
	{ "valuelen", FLDT_UINT32D, OI(LVOFF(valuelen)),
	  attr_leaf_name_remote_count, FLD_COUNT, TYP_NONE },
	{ "namelen", FLDT_UINT8D, OI(LVOFF(namelen)),
	  attr_leaf_name_remote_count, FLD_COUNT, TYP_NONE },
	{ "name", FLDT_CHARNS, OI(LVOFF(name)),
	  attr_leaf_name_remote_name_count, FLD_COUNT, TYP_NONE },
	{ NULL }
};

#define	EOFF(f)	bitize(offsetof(xfs_da_node_entry_t, f))
const field_t	attr_node_entry_flds[] = {
	{ "hashval", FLDT_UINT32X, OI(EOFF(hashval)), C1, 0, TYP_NONE },
	{ "before", FLDT_ATTRBLOCK, OI(EOFF(before)), C1, 0, TYP_ATTR },
	{ NULL }
};

#define	HOFF(f)	bitize(offsetof(xfs_da_node_hdr_t, f))
const field_t	attr_node_hdr_flds[] = {
	{ "info", FLDT_ATTR_BLKINFO, OI(HOFF(info)), C1, 0, TYP_NONE },
	{ "count", FLDT_UINT16D, OI(HOFF(count)), C1, 0, TYP_NONE },
	{ "level", FLDT_UINT16D, OI(HOFF(level)), C1, 0, TYP_NONE },
	{ NULL }
};

/*ARGSUSED*/
static int
attr_leaf_entries_count(
	void			*obj,
	int			startoff)
{
	xfs_attr_leafblock_t	*block;
	
	assert(startoff == 0);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	return block->hdr.count;
}

/*ARGSUSED*/
static int
attr_leaf_hdr_count(
	void			*obj,
	int			startoff)
{
	xfs_attr_leafblock_t	*block;
	
	assert(startoff == 0);
	block = obj;
	return block->hdr.info.magic == XFS_ATTR_LEAF_MAGIC;
}

static int
attr_leaf_name_local_count(
	void			*obj,
	int			startoff)
{
	xfs_attr_leafblock_t	*block;
	xfs_attr_leaf_entry_t	*e;
	int			i;
	int			off;

	assert(bitoffs(startoff) == 0);
	off = byteize(startoff);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	for (i = 0; i < block->hdr.count; i++) {
		e = &block->entries[i];
		if (e->nameidx == off)
			return (e->flags & XFS_ATTR_LOCAL) != 0;
	}
	return 0;
}

static int
attr_leaf_name_local_name_count(
	void				*obj,
	int				startoff)
{
	xfs_attr_leafblock_t		*block;
	xfs_attr_leaf_entry_t		*e;
	int				i;
	xfs_attr_leaf_name_local_t	*l;
	int				off;

	assert(bitoffs(startoff) == 0);
	off = byteize(startoff);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	for (i = 0; i < block->hdr.count; i++) {
		e = &block->entries[i];
		if (e->nameidx == off) {
			if (e->flags & XFS_ATTR_LOCAL) {
				l = XFS_ATTR_LEAF_NAME_LOCAL(block, i);
				return l->namelen;
			} else
				return 0;
		}
	}
	return 0;
}

static int
attr_leaf_name_local_value_count(
	void				*obj,
	int				startoff)
{
	xfs_attr_leafblock_t		*block;
	xfs_attr_leaf_entry_t		*e;
	int				i;
	xfs_attr_leaf_name_local_t	*l;
	int				off;

	assert(bitoffs(startoff) == 0);
	off = byteize(startoff);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	for (i = 0; i < block->hdr.count; i++) {
		e = &block->entries[i];
		if (e->nameidx == off) {
			if (e->flags & XFS_ATTR_LOCAL) {
				l = XFS_ATTR_LEAF_NAME_LOCAL(block, i);
				return l->valuelen;
			} else
				return 0;
		}
	}
	return 0;
}

/*ARGSUSED*/
static int
attr_leaf_name_local_value_offset(
	void				*obj,
	int				startoff,
	int				idx)
{
	xfs_attr_leafblock_t		*block;
	xfs_attr_leaf_name_local_t	*l;
	char				*vp;

	assert(bitoffs(startoff) == 0);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	l = XFS_ATTR_LEAF_NAME_LOCAL(block, idx);
	vp = (char *)&l->nameval[l->namelen];
	return (int)bitize(vp - (char *)l);
}

static int
attr_leaf_name_remote_count(
	void			*obj,
	int			startoff)
{
	xfs_attr_leafblock_t	*block;
	xfs_attr_leaf_entry_t	*e;
	int			i;
	int			off;

	assert(bitoffs(startoff) == 0);
	off = byteize(startoff);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	for (i = 0; i < block->hdr.count; i++) {
		e = &block->entries[i];
		if (e->nameidx == off)
			return (e->flags & XFS_ATTR_LOCAL) == 0;
	}
	return 0;
}

static int
attr_leaf_name_remote_name_count(
	void				*obj,
	int				startoff)
{
	xfs_attr_leafblock_t		*block;
	xfs_attr_leaf_entry_t		*e;
	int				i;
	int				off;
	xfs_attr_leaf_name_remote_t	*r;

	assert(bitoffs(startoff) == 0);
	off = byteize(startoff);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	for (i = 0; i < block->hdr.count; i++) {
		e = &block->entries[i];
		if (e->nameidx == off) {
			if (!(e->flags & XFS_ATTR_LOCAL)) {
				r = XFS_ATTR_LEAF_NAME_REMOTE(block, i);
				return r->namelen;
			} else
				return 0;
		}
	}
	return 0;
}

/*ARGSUSED*/
int
attr_leaf_name_size(
	void				*obj,
	int				startoff,
	int				idx)
{
	xfs_attr_leafblock_t		*block;
	xfs_attr_leaf_entry_t		*e;
	xfs_attr_leaf_name_local_t	*l;
	xfs_attr_leaf_name_remote_t	*r;

	assert(startoff == 0);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	e = &block->entries[idx];
	if (e->flags & XFS_ATTR_LOCAL) {
		l = XFS_ATTR_LEAF_NAME_LOCAL(block, idx);
		return (int)bitize(XFS_ATTR_LEAF_ENTSIZE_LOCAL(l->namelen,
								l->valuelen));
	} else {
		r = XFS_ATTR_LEAF_NAME_REMOTE(block, idx);
		return (int)bitize(XFS_ATTR_LEAF_ENTSIZE_REMOTE(r->namelen));
	}
}

/*ARGSUSED*/
static int
attr_leaf_nvlist_count(
	void			*obj,
	int			startoff)
{
	xfs_attr_leafblock_t	*block;

	assert(startoff == 0);
	block = obj;
	if (block->hdr.info.magic != XFS_ATTR_LEAF_MAGIC)
		return 0;
	return block->hdr.count;
}

/*ARGSUSED*/
static int
attr_leaf_nvlist_offset(
	void			*obj,
	int			startoff,
	int			idx)
{
	xfs_attr_leafblock_t	*block;
	xfs_attr_leaf_entry_t	*e;

	assert(startoff == 0);
	block = obj;
	e = &block->entries[idx];
	return bitize(e->nameidx);
}

/*ARGSUSED*/
static int
attr_node_btree_count(
	void			*obj,
	int			startoff)
{
	xfs_da_intnode_t	*block;

	assert(startoff == 0);		/* this is a base structure */
	block = obj;
	if (block->hdr.info.magic != XFS_DA_NODE_MAGIC)
		return 0;
	return block->hdr.count;
}

/*ARGSUSED*/
static int
attr_node_hdr_count(
	void			*obj,
	int			startoff)
{
	xfs_da_intnode_t	*block;
	
	assert(startoff == 0);
	block = obj;
	return block->hdr.info.magic == XFS_DA_NODE_MAGIC;
}

/*ARGSUSED*/
int
attr_size(
	void	*obj,
	int	startoff,
	int	idx)
{
	return bitize(mp->m_sb.sb_blocksize);
}
